src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,215 @@
+#include <clp/ir/LogEvent.hpp>
+#include <clp/ir/LogEventDeserializer.hpp>
+#include <clp/ir/types.hpp>
+#include <clp/streaming_compression/zstd/Decompressor.hpp>
+#include <clp/TimestampPattern.hpp>
+#include <clp/TraceableException.hpp>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
+#include <span>
+#include <system_error>
+#include <utility>
+
+#include <spdlog/spdlog.h>
+
+#include "ClpJsException.hpp"
+#include "constants.hpp"
+
+using namespace std::literals::string_literals;
+
+// Constants
+namespace {
+constexpr size_t cDefaultNumCharsPerMessage{512};
+constexpr size_t cDefaultNumLogEvents{500'000};
+constexpr size_t cFullRangeEndIdx{0};
+constexpr size_t cLogLevelNone{0};
+}  // namespace
+
+auto ClpIrV1Decoder::create(emscripten::val const& data_array) -> std::unique_ptr<ClpIrV1Decoder> {
+    auto length{data_array["length"].as<size_t>()};
+    SPDLOG_INFO("ClpIrV1Decoder::ClpIrV1Decoder() got buffer of length={}", length);
+
+    auto data_buffer{std::make_unique<char const[]>(length)};
+    emscripten::val::module_property("HEAPU8")
+            .call<void>("set", data_array, reinterpret_cast<uintptr_t>(data_buffer.get()));
+
+    auto zstd_decompressor{std::make_shared<clp::streaming_compression::zstd::Decompressor>()};
+    zstd_decompressor->open(data_buffer.get(), length);
+
+    bool is_four_bytes_encoding{true};
+    if (auto const err{
+                clp::ffi::ir_stream::get_encoding_type(*zstd_decompressor, is_four_bytes_encoding)
+        };
+        clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success != err)
+    {
+        SPDLOG_CRITICAL("Failed to decode encoding type, err={}", err);
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decode encoding type.."
+        );
+    }
+
+    if (false == is_four_bytes_encoding) {
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_Unsupported,
+                __FILENAME__,
+                __LINE__,
+                "Is not four byte encoding."
+        );
+    }
+    auto result{clp::ir::LogEventDeserializer<clp::ir::four_byte_encoded_variable_t>::create(
+            *zstd_decompressor
+    )};
+    if (result.has_error()) {
+        auto const error_code{result.error()};
+        SPDLOG_CRITICAL(
+                "Failed to decompress: {}:{}",
+                error_code.category().name(),
+                error_code.message()
+        );
+        throw ClpJsException(
+                clp::ErrorCode::ErrorCode_MetadataCorrupted,
+                __FILENAME__,
+                __LINE__,
+                "Failed to decompress"
+        );
+    }
+
+    std::unique_ptr<ClpIrV1Decoder> ptr(
+            new ClpIrV1Decoder(std::move(data_buffer), zstd_decompressor, std::move(result.value()))
+    );
+    return ptr;
+}
+
+auto ClpIrV1Decoder::get_estimated_num_events() -> size_t const {
```

Comment:
```
Clang-Tidy: Return type 'const size_t' (aka 'const unsigned long') is 'const'-qualified at the top level, which may reduce code readability without improving const correctness 
```
@LinZhihao-723 Do you have an idea about how to fix that?

Reply:
```suggestion
auto ClpIrV1Decoder::get_estimated_num_events() const -> size_t {
```